finetune/target_models/sequence_labeling.py

```diff
@@ -265,6 +306,45 @@ def predict(self, X, per_token=False, context=None, **kwargs):
             subtoken_predictions=self.config.subtoken_predictions,
         )
 
+        if context is not None and self.config.seq_labeling_2d_spans:
+            doc_annotations_out = []
+            for annotation, sample_context in zip(doc_annotations, context):
+                for single_label in annotation:
+                    single_label["contexts"] = []
+                    for cont in sample_context:
+                        if sequences_overlap(cont, single_label):
+                            single_label["contexts"].append(cont) # cannot break here because one label has multiple contexts.
+                    if not single_label["contexts"] and single_label["text"].strip():
+                        raise FinetuneError("Context does not cover char idx: {} to {}".format(single_label["start"], single_label["end"]))
+                    single_label["start_segments"] = [single_label.pop("start")]
+                    single_label["end_segments"] = [single_label.pop("end")]
+                    single_label["text_segments"] = [single_label.pop("text")]
+                    single_label["confidence_segments"] = [single_label.pop("confidence")]
+      
+                for _ in range(len(annotation)):
+                    len_at_start = len(annotation)
+                    annotation_out = [annotation[0]]
+                    annotation = annotation[1:]
+                    for a in annotation:
+                        for out in annotation_out:
+                            if a["label"] == out["label"] and _context_same_cluster(a["contexts"], out["contexts"], thresholds=self.config.predict_span_threshold):
+                                out["start_segments"] += a["start_segments"]
+                                out["end_segments"] += a["end_segments"]
+                                out["text_segments"] += a["text_segments"]
+                                out["contexts"] += a["contexts"]
+                                out["confidence_segments"] += a["confidence_segments"]
+                                break
+                        else:
+                            annotation_out.append(a)
+                    annotation = annotation_out
+                    if len(annotation_out) == len_at_start:
+                        break
+                    
```

Comment:
I feel like this makes some assumptions about order / may not necessarily merge all contiguous tokens?